### PR TITLE
Add storage wrapper

### PR DESCRIPTION
**Pull request by VicDeo**
Fixes https://github.com/owncloud/files_antivirus/issues/15

### CODE REVIEW

**Comment by VicDeo** [comment]
@DeepDiver1975 I can't trigger `file_put_contents` on storage. 
According to the trace uploading a new file uses `\OC\Files\Filesystem::fromTmpFile` which calls  `file_put_contents` on view. https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L480

Storage doen't involved in this process

**Comment by DeepDiver1975** [comment]
the voew operations use basicOperation, which will redirect the calls to the storage - https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L954

File: lib/avirwrapper.php [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+			
+			$content = new Content($data, $this->storage);
+			
+			$scanner = new Scanner($appConfig, $l10n);
+			$status = $scanner->scan($content);
+		}
+		return $this->storage->file_put_contents($path, $data);
+	}
+	
+	public static function setupWrapper() {
```
Comment (icewind1991) [inline lib/avirwrapper.php]: please no statics

Comment (VicDeo) [inline lib/avirwrapper.php]: @icewind1991 Ok, so what is preferable to setup a wrapper? This is used as a callback to instantiate this class during FS setup

**Comment by VicDeo** [comment]
@DeepDiver1975 basicOperation is not called  'data' is resource and execution goes through another branch
trace

**Comment by VicDeo** [comment]
A real working horse in this case is `list ($count, $result) = \OC_Helper::streamCopy($data, $target);`

https://github.com/owncloud/core/blob/master/lib/private/files/view.php#L498
and that's the answer to my question :(

**Comment by DeepDiver1975** [comment]
> and that's the answer to my question :(

so - file_put_contents is never called - but only fopen - which means you need to come up with the stream wrapper approach.

**Comment by DeepDiver1975** [comment]
@icewind1991 looks like we should merge your stream wrappers into 3rdparty and reuse them here?
https://github.com/owncloud/3rdparty/pull/164

**Comment by DeepDiver1975** [comment]
@VicDeo I just merged owncloud/3rdparty#164 - please make use of Icewind\Streams to implement the stream wrapper approach - THX

**Comment by DeepDiver1975** [comment]
Would be great to have something to test by the end of this week - THX

**Comment by VicDeo** [comment]
@DeepDiver1975 ok, on it

**Comment by VicDeo** [comment]
yes, I' aware of this:

```
{
"reqId":"UT3gYXSsuUHQueTkwKh6",
"remoteAddr":"::1",
"app":"webdav",
"message":"Exception: {Message:expected filesize 2605 got 0",
Code:0,
Trace:

#0 /master/3rdparty/sabre/dav/lib/DAV/Server.php(1087): OC\Connector\Sabre\File->put(Resource id #264)

#1 /master/3rdparty/sabre/dav/lib/DAV/CorePlugin.php(501): Sabre\DAV\Server->updateFile('alf.png', Resource id #264, NULL)

#2 [internal function]: Sabre\DAV\CorePlugin->httpPut(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))

#3 /master/3rdparty/sabre/event/lib/EventEmitterTrait.php(105): call_user_func_array(Array, Array)

#4 /master/3rdparty/sabre/dav/lib/DAV/Server.php(469): Sabre\Event\EventEmitter->emit('method:PUT', Array)

#5 /master/3rdparty/sabre/dav/lib/DAV/Server.php(254): Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))

#6 /master/apps/files/appinfo/remote.php(71): Sabre\DAV\Server->exec()

#7 /master/remote.php(58): require_once('/home/deo/publi...')

#8{main}File:/master/lib/private/connector/sabre/file.php,Line:130}","level":4,"time":"2015-03-11T14:49:28+00:00","method":"PUT","url":"/~deo/oc-tmp/remote.php/webdav/alf.png"}
```

File: lib/avirwrapper.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	 */
+	public function fopen($path, $mode){
+		$stream = $this->storage->fopen($path, $mode);
+		if (is_resource($stream)
+				&& $this->file_exists($path) 
+				&& $this->is_file($path) 
+				&& $this->isWritingMode($mode)
+		) {
+			try {
+				$application = new \OCA\Files_Antivirus\AppInfo\Application();
```
Comment (icewind1991) [inline lib/avirwrapper.php]: inject this in the wrapper constructor?

**Comment by VicDeo** [comment]
Ok, got it working.
Uploading 03labfs1t1348649414.jpg to `/master/remote.php/webdav/03labfs1t1348649414.jpg':
Progress: [=============================>] 100.0% of 270607 bytes succeeded.
dav:/~deo/oc-tmp/remote.php/webdav/> put testvirus.txt 
Uploading testvirus.txt to `/master/remote.php/webdav/testvirus.txt':
Progress: [=============================>] 100.0% of 68 bytes failed:
415 Unsupported Media Type

**Comment by DeepDiver1975** [comment]
> Ok, got it working.

awesome - you rock!

**Comment by DeepDiver1975** [comment]
@VicDeo what's the status on this? THX

The error on travis:
````
.PHP Fatal error: Call to a member function initScanner() on a non-object in /home/travis/build/owncloud/core/apps/files_antivirus/lib/scanner.php on line 87

PHP Stack trace:

PHP 1. {main}() /home/travis/.phpenv/versions/5.4.37/bin/phpunit:0

PHP 2. PHPUnit_TextUI_Command::main() /home/travis/.phpenv/versions/5.4.37/bin/phpunit:722

PHP 3. PHPUnit_TextUI_Command->run() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/Command.php:104

PHP 4. PHPUnit_TextUI_TestRunner->doRun() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/Command.php:152

PHP 5. PHPUnit_Framework_TestSuite->run() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/TextUI/TestRunner.php:398

PHP 6. PHPUnit_Framework_TestSuite->run() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestSuite.php:716

PHP 7. PHPUnit_Framework_TestCase->run() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestSuite.php:716

PHP 8. PHPUnit_Framework_TestResult->run() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestCase.php:693

PHP 9. PHP_Invoker->invoke() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestResult.php:607

PHP 10. call_user_func_array:{phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/php-invoker/Invoker.php:93}() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/php-invoker/Invoker.php:93

PHP 11. PHPUnit_Framework_TestCase->runBare() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/php-invoker/Invoker.php:93

PHP 12. PHPUnit_Framework_TestCase->runTest() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestCase.php:737

PHP 13. ReflectionMethod->invokeArgs() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestCase.php:860

PHP 14. Test_Files_Antivirus_ScannerTest->testBackgroundScan() phar:///home/travis/.phpenv/versions/5.4.37/bin/phpunit/phpunit/Framework/TestCase.php:860

PHP 15. OCA\Files_Antivirus\BackgroundScanner->run() /home/travis/build/owncloud/core/apps/files_antivirus/tests/scannertest.php:55

PHP 16. OCA\Files_Antivirus\Scanner->scan() /home/travis/build/owncloud/core/apps/files_antivirus/lib/backgroundscanner.php:77

File: lib/avirwrapper.php (hunk 2 above)
Comment (VicDeo) [inline lib/avirwrapper.php]: @icewind1991 sadly I can't:
>Declaration of OCA\Files_Antivirus\AvirWrapper::__construct() must be compatible with OCP\Files\Storage::__construct($parameters)

Comment (DeepDiver1975) [inline lib/avirwrapper.php]: @VicDeo well adopting the ctor of AvirWrapper accordingly should not be an issue - right?

Comment (VicDeo) [inline lib/avirwrapper.php]: @DeepDiver1975 for the test to fail?  Definitely no. 
There is something wrong with reading `av_mode` config option there

Comment (icewind1991) [inline lib/avirwrapper.php]: You can pass it as part of the arguments array

Comment (VicDeo) [inline lib/avirwrapper.php]: @icewind1991 done

**Comment by VicDeo** [comment]
@DeepDiver1975 Tests are passing 
Oracle says: `adduser: The group `dba' does not exist.`

**Comment by VicDeo** [comment]
@icewind1991 moved wrapper setup into application

**Comment by DeepDiver1975** [comment]
> Oracle says: adduser: The groupdba' does not exist.`

because oracel server setup did fail - https://travis-ci.org/owncloud/files_antivirus/jobs/54608709#L348

File: appinfo/app.php [hunk 3]
```diff
@@ -25,7 +25,7 @@
 OCP\BackgroundJob::AddRegularTask('OCA\Files_Antivirus\Cron\Task', 'run');
 
 $app = new \OCA\Files_Antivirus\AppInfo\Application();
-$app->getContainer()->query('FilesystemHooks')->register();
+OCP\Util::connectHook('OC_Filesystem', 'setup', '\OCA\Files_Antivirus\AppInfo\Application', 'setupWrapper');
```
Comment (icewind1991) [inline appinfo/app.php]:
```php
OCP\Util::connectHook('OC_Filesystem', 'setup', $app, 'setupWrapper');
```

And make the method non-static

File: appinfo/application.php [hunk 4]
```diff
@@ -77,4 +78,27 @@ public function __construct (array $urlParams = array()) {
...
+	/**
+	 * Wrapper for local storages
+	 */
+	public static function setupWrapper(){
+		\OC\Files\Filesystem::addStorageWrapper('oc_avir', function ($mountPoint, $storage) {
+			/**
+			 * @var \OC\Files\Storage\Storage $storage
+			 */
+			if ($storage instanceof \OC\Files\Storage\Storage && $storage->isLocal()) {
+				$application = new self();
```
Comment (icewind1991) [inline appinfo/application.php]: once this is made non static (see comment above) you can just use `$this`

**Comment by VicDeo** [comment]
@icewind1991 done

File: lib/avirwrapper.php [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+					function () use ($scanner, $storage, $path) {
+						$status = $scanner->completeAsyncScan();
+						if ($status->getNumericStatus() == \OCA\Files_Antivirus\Status::SCANRESULT_INFECTED){
+							$storage->unlink($path);
+							throw new InvalidContentException($status->getDetails());
+						}
+					}
+				);
+			} catch (\Exception $e){
+
```
Comment (DeepDiver1975) [inline lib/avirwrapper.php]: Log the exception?

File: lib/avirwrapper.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	/**
+	 * Scan content on-the-fly
+	 * @param string $path
+	 * @param string $data
+	 * @return bool
+	 */
+	public function file_put_contents($path, $data){
+	/*	if (!$this->storage->is_dir($path)) {
```
Comment (DeepDiver1975) [inline lib/avirwrapper.php]: no scanning on file_put_contents ? - At least remove commented code block

Comment (VicDeo) [inline lib/avirwrapper.php]: @DeepDiver1975 In general it would be nice to have both. 
The current *dbprefix*files_antivirus table uses the time of the last scan which is pretty useless and doesn't store etag in order to prevent the file being scanned twice. 

But I don't want put these changes in this PR. So OK, this method will be removed for now.

File: lib/avirwrapper.php (hunk 5 above)
Comment (VicDeo) [inline lib/avirwrapper.php]: @DeepDiver1975 oh, I might be too old-fashioned to inject logger into constructor through array value. Just dislike the idea... 
ok, will do.

Comment (VicDeo) [inline lib/avirwrapper.php]: offtop
Just to be clear. I  more like the idea of an app-specific Singleton that has logger, l10n and all other service stuff defined via an interface. 
Something that is already mocked during tests instead of injecting the same service object instances that are not related to business logic here and there.

**Comment by DeepDiver1975** [comment]
@icewind1991 any further objections? - I'm :+1: to get this in

File: lib/avirwrapper.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+	 * Asyncronously scan data that are written to the file
+	 * @param string $path
+	 * @param string $mode
+	 * @return resource | bool
+	 */
+	public function fopen($path, $mode){
+		$stream = $this->storage->fopen($path, $mode);
+		if (is_resource($stream) && $this->isWritingMode($mode)) {
+			try {
+				$storage = $this;
```
Comment (icewind1991) [inline lib/avirwrapper.php]: no need to do this since php 5.4 you can just use `$this` from your closure

File: lib/content.php [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	protected $currentPosition = 0;
+	
+	protected $chunkSize;
+	
+	public function __construct($content, $storage){
+		$this->content = $content;
+		$this->storage = $storage;
+		$application = new \OCA\Files_Antivirus\AppInfo\Application();
+		$config = $application->getContainer()->query('AppConfig');
+		$this->chunkSize = $config->getAvChunkSize();
```
Comment (icewind1991) [inline lib/content.php]: either the config or chunksize should be injected if possible

File: lib/scanner.php [hunk 9]
```diff
@@ -24,7 +24,6 @@
 namespace OCA\Files_Antivirus;
 
 use OCP\IL10N;
-use OCA\Files_Antivirus\Item;
 
 class Scanner {
```
Comment (icewind1991) [inline lib/scanner.php]: If I understand correctly this class functions both as factory/wrapper and base class.

I would make this class either and interface or an abstract class and put the factory logic in a new class

**Comment by VicDeo** [comment]
@icewind1991 done

**Comment by icewind1991** [comment]
:+1: nice work

**Comment by butonic** [comment]
@dragotin How does the desktop client handle 415 Unsupported Media Type?
